### PR TITLE
CWE-1004: Sensitive cookie without HttpOnly

### PR DESCRIPTION
**Pull request by edvraa**
This PR also fixes `secure` flag handling for `cookie-session` because its options are flat like `session({ name: 'a', secure: true})`, not `session({ name: 'a', cookie: { secure: true }})`.
It also fixes `InsecureSetCookieHeader`:
- Cookie attributes are in fact case insensitive.
- `res.setHeader("Set-Cookie", "auth=ninja");` (not array) was not supported.
- In the array case `res.setHeader("Set-Cookie", ["auth=ninja; secure", "token=javascript"]);` if any of the cookies had `secure` set the other one was not flagged.

The HttpOnly-less cookies are flagged only if the name is likely related to authentication: `auth`, `session`, `login`, etc. Anti-forgery tokens are explicitly excluded to reduce False Positives as it is common practice to make them available to JavaScript https://en.wikipedia.org/wiki/Cross-site_request_forgery#Cookie-to-header_token

### CODE REVIEW

**Comment by esbena** [comment]
Thanks!
I know of two concrete cases where this query is relevant:

- A: [Application with seeded vulnerabilities](https://github.com/SecureSkyTechnology/BadLibrary/blob/13d4466cc9745cf6ee733431e61e029a71536c20/src/waf.js#L193)
- B: [Diff](https://github.com/fastify/fastify-csrf/commit/3c9de36e9e73ce0eda9207f84f2ac0243e1f5253) for CVE-2020-28482

Could you perhaps first check if the query produces results that you are satisfied with in both cases?

**Comment by edvraa** [comment]
Hello,
The query wouldn't flag both issues for multiple reasons:
- A:
  - The `config` value `httponly` is [set to true](https://github.com/SecureSkyTechnology/BadLibrary/blob/13d4466cc9745cf6ee733431e61e029a71536c20/src/waf.js#L13) 
  - Even if it was indeterminate on [line 193](https://github.com/SecureSkyTechnology/BadLibrary/blob/13d4466cc9745cf6ee733431e61e029a71536c20/src/waf.js#L193): `${config.httponly ? 'httponly' : ''` I thought CodeQL queries are designed to be low false positives even at cost of possible false negatives.
  - Differently from the other sinks it is a simple regex in `InsecureSetCookieHeader` case. It wouldn't catch the string interpolation expression. I think I could change that into taint tracking where the source would be a string containing any `httponly` word. Do you think that could work?
- B:
  - Fastify-csrf uses fastify-cookie library. The query doesn't support it. It looks like the library isn't very popular, so I don't know if it is worth implementing support for it.
  - In my opinion if there is a XSS vulnerability (to read the cookie from injected JS) it is a lot worse than CSRF. I don't know why Fastify decided to request a CVE for it. Overall it is arguable if CSRF cookies have to be httponly or not and probably depends on project. I wanted to avoid unnecessary false positives, so try to exclude them. But I can change to include them. Another option is to make two queries: with CSRF cookies excluded and not. What do you think?

**Review by esbena — CHANGES_REQUESTED** [review]
This generally LGTM. I look forward to the results.

Big thing: 
Do we really need to move `InsecureCookie` out of `/experimental` though? The review requirements for that move are much more involved...

> The query wouldn't flag both issues for multiple reasons: ...

Ok. Thanks for clarifying your choices. It sounds like query potentially could handle those cases with some additional work.

File: javascript/ql/src/semmle/javascript/security/InsecureCookie.qll [hunk 1]
```diff
@@ -105,14 +176,42 @@ module Cookie {
...
-        getCookieOptionsArgument().mayHaveStringValue(s) and
-        s.regexpMatch("(.*;)?\\s*secure.*")
+      // The default is `false`.
+      forall(DataFlow::Node n | n = getCookieOptionsArgument() |
+        exists(string s |
+          n.mayHaveStringValue(s) and
+          s.regexpMatch("(?i).*;\\s*secure\\s*;?.*$")
+        )
+      )
+    }
```
[the remark this answers is not in the document]
Comment (edvraa) [reply javascript/ql/src/semmle/javascript/security/InsecureCookie.qll]: - In array case multiple cookies are set. Each element has separate attributes.
- Header value always starts with cookie name=value
As I wrote in PR description it also fixes `Secure` attribute matching. I have added more comments in the code.

File: javascript/ql/src/semmle/javascript/security/InsecureCookie.qll [hunk 2]
```diff
@@ -105,14 +176,42 @@ module Cookie {
...
+            regexpMatchAuth(s.regexpCapture("\\s*([^=\\s]*)\\s*=.*", 1))
+          )
+        )
+      )
+    }
+
+    override predicate isHttpOnly() {
+      // A cookie is httpOnly if the 'httpOnly' flag is specified in the cookie definition.
+      // The default is `false`.
+      forall(DataFlow::Node n | n = getCookieOptionsArgument() |
```
[the remark this answers is not in the document]
Comment (edvraa) [reply javascript/ql/src/semmle/javascript/security/InsecureCookie.qll]: No, each array element has separate attributes.

**Review by esbena — COMMENTED** [review]
Thanks. Especially for the clarification about the array signature. 

I have a few remaining nits regarding the documentation. If you agree, and just commit my suggestions, then we are ready to merge this PR.

**Comment by esbena** [comment]
The alert message changed, so the tests are failing now. Sorry about not warning you about that. Can you bump the expected output of CookieWithoutHttpOnly.qlref?

**Comment by edvraa** [comment]
Sure

**Review by esbena — APPROVED** [review]

**Comment by esbena** [comment]
Thank you for the contribution.

**Pull request merged**